flink-cdc-runtime/src/main/java/com/ververica/cdc/runtime/operators/schema/coordinator/SchemaManager.java

```diff
@@ -16,10 +16,154 @@
 
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+
+import com.ververica.cdc.common.event.SchemaChangeEvent;
+import com.ververica.cdc.common.event.TableId;
+import com.ververica.cdc.common.serializer.ListSerializer;
+import com.ververica.cdc.common.serializer.MapSerializer;
+import com.ververica.cdc.common.serializer.TableIdSerializer;
+import com.ververica.cdc.common.serializer.event.SchemaChangeEventSerializer;
+import com.ververica.cdc.common.sink.MetadataApplier;
+import com.ververica.cdc.runtime.operators.schema.event.ReleaseUpstreamEvent;
+import com.ververica.cdc.runtime.operators.schema.event.SchemaChangeRequest;
+import com.ververica.cdc.runtime.operators.schema.event.SchemaChangeResponse;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
 /** A manager to manages schema changes. */
 public class SchemaManager {
     private static final Logger LOG = LoggerFactory.getLogger(SchemaManager.class);
+
+    /** A serializer to serialize and deserialize checkpoint of the coordinator. */
+    private static final MapSerializer<TableId, List<SchemaChangeEvent>> STATE_SERIALIZER =
+            new MapSerializer<>(
+                    TableIdSerializer.INSTANCE,
+                    new ListSerializer(SchemaChangeEventSerializer.INSTANCE));
+
+    /** The MetadataApplier for evey sink. */
+    private final Map<TableId, List<MetadataApplier>> metadataAppliers;
+    /** All active sink writers. */
+    private final Set<Integer> activeSinkWriters;
+
+    /**
+     * Not applied SchemaChangeRequest before receiving all flush success events for its table from
+     * sink writers.
+     */
+    private Map<TableId, List<SchemaChangeEvent>> pendingRequests;
+    /** Key => tableId, val => sink writers which have sent flush success events for the table. */
+    private final Map<TableId, Set<Integer>> nonFlushedSinkWriters;
+
+    public SchemaManager(Map<TableId, List<MetadataApplier>> metadataAppliers) {
+        this.metadataAppliers = metadataAppliers;
+        this.activeSinkWriters = new HashSet<>();
+        this.pendingRequests = new HashMap<>();
+        this.nonFlushedSinkWriters = new HashMap<>();
+    }
+
+    /**
+     * Apply the schema change to the external system.
+     *
+     * @param tableId the table need to change schema
+     * @param changeEvents the schema changes
+     */
+    private void applySchemaChange(TableId tableId, List<SchemaChangeEvent> changeEvents) {
+        List<MetadataApplier> appliers = metadataAppliers.get(tableId);
+        if (appliers == null || appliers.isEmpty()) {
+            LOG.warn("There is no MetadataApplier for table {}.", tableId);
```

Comment:
I think timely UnsupportedException is better here,  because the following data change event would crash at sink end after a schema change happened